instantout/manager.go

```diff
@@ -0,0 +1,182 @@
+package instantout
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/lightninglabs/loop/instantout/reservation"
+	"github.com/lightningnetwork/lnd/lntypes"
+)
+
+var (
+	defaultStateWaitTime = 30 * time.Second
+	ErrSwapDoesNotExist  = errors.New("swap does not exist")
+)
+
+// Manager manages the instantout state machines.
+type Manager struct {
+	// cfg contains all the services that the reservation manager needs to
+	// operate.
+	cfg *Config
+
+	// activeInstantOuts contains all the active instantouts.
+	activeInstantOuts map[lntypes.Hash]*FSM
+
+	// currentHeight stores the currently best known block height.
+	currentHeight int32
+
+	// blockEpochChan receives new block heights.
+	blockEpochChan chan int32
+
+	runCtx context.Context
+
+	sync.Mutex
+}
+
+// NewInstantOutManager creates a new instantout manager.
+func NewInstantOutManager(cfg *Config) *Manager {
+	return &Manager{
+		cfg:               cfg,
+		activeInstantOuts: make(map[lntypes.Hash]*FSM),
+		blockEpochChan:    make(chan int32),
+	}
+}
+
+// Run runs the reservation manager.
+func (i *Manager) Run(ctx context.Context, height int32) error {
+	log.Debugf("Starting reservation manager")
+
+	runCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	i.Lock()
+	i.runCtx = runCtx
+	i.currentHeight = height
+	i.Unlock()
+
+	err := i.RecoverInstantOuts(runCtx)
+	if err != nil {
+		return err
+	}
+
+	newBlockChan, newBlockErrChan, err := i.cfg.ChainNotifier.
+		RegisterBlockEpochNtfn(ctx)
+	if err != nil {
+		return err
+	}
+
+	for {
+		select {
+		case <-runCtx.Done():
+			log.Debugf("Stopping reservation manager")
+			return nil
+		case height := <-newBlockChan:
+			i.Lock()
+			i.currentHeight = height
+			i.Unlock()
+
+		case err := <-newBlockErrChan:
+			return err
+		}
+	}
+}
+
+// RecoverInstantOuts recovers all the active instantouts from the database.
+func (i *Manager) RecoverInstantOuts(ctx context.Context) error {
+	// Fetch all the active instantouts from the database.
+	activeInstantOuts, err := i.cfg.Store.ListInstantLoopOuts(ctx)
+	if err != nil {
+		return err
+	}
+
+	for _, instantOut := range activeInstantOuts {
+		if isFinalState(instantOut.State) {
+			continue
+		}
+
+		log.Debugf("Recovering instantout %v",
+			instantOut.SwapHash)
+
+		instantOutFSM, err := NewFSMFromInstantOut(
+			ctx, i.cfg, instantOut,
+		)
+		if err != nil {
+			return err
+		}
+
+		i.activeInstantOuts[instantOut.SwapHash] = instantOutFSM
```

Comment:
don't think this can run when `GetActiveInstantOut` runs, but maybe we still lock this.

Reply:
It's a valid concern, by making the recover function private and adding an init channel we can make sure init is done and can avoid the locking. Obviously adding the lock doesn't hurt.

Alternatively we could drive all calls through the runloop so no locks are needed.